### PR TITLE
Registry API change

### PR DESCRIPTION
**Pull request by philipktlin**
Replace RegistryMiddleware by RegistryController using ASP.Net MVC Controller, and enable On-behalf-of API endpoint as well.

### CODE REVIEW

File: builds/misc/images.yaml [hunk 1]
```diff
@@ -184,26 +184,26 @@ jobs:
 ################################################################################
 # linux - Rust amd64
 ################################################################################
-  - template: templates/image-linux-rust-jobs.yaml
-    parameters:
-      archName: amd64
-      targetArch: x86_64
+  # - template: templates/image-linux-rust-jobs.yaml
```
Comment (philipktlin) [inline builds/misc/images.yaml:187]: Ignore all changes in this file.  This is just to unblock image build for testing.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/RegistryApiClient.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        public async Task<HttpResponseMessage> DeleteModuleOnBehalfOfAsync(string actorEdgeDeviceId, DeleteModuleOnBehalfOfData requestData)
+        {
+            var requesteUri = new Uri(this.baseUri, string.Format(CultureInfo.InvariantCulture, DeleteModuleOnBehalfOfUriTemplate, actorEdgeDeviceId, ApiVersion));
+            var response = await this.SendRequestAsync(requesteUri, actorEdgeDeviceId, HttpMethod.Post, requestData);
+            return response;
+        }
+
+        async Task<HttpResponseMessage> SendRequestAsync<T>(Uri uri, string actorEdgeDeviceId, HttpMethod method, T payload, string etag = null)
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/RegistryApiClient.cs:66]: Will remove "actorEdgeDeviceId"

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/ControllerHelper.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Controllers
+{
+    using System;
+
+    static class ControllerHelper
+    {
+        internal static bool TryGetTargetDeviceId(string authChain, out string targetDeviceId)
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/ControllerHelper.cs:8]: TODO: will use it in DeviceScopeController.cs

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/HttpRequestAuthenticator.cs [hunk 4]
```diff
@@ -39,6 +39,7 @@ public async Task<HttpAuthResult> AuthenticateAsync(string deviceId, Option<stri
 
             IClientCredentials clientCredentials;
             X509Certificate2 clientCertificate = await context.Connection.GetClientCertificateAsync();
+            Events.AuthenticationDebug($"deviceId: {deviceId}, moduleId: {moduleId.GetOrElse(string.Empty)}, client certificate: {clientCertificate != null}");
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/HttpRequestAuthenticator.cs:42]: ignore all changes in this file, those are used for debugging.  Will remove before commit.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/ControllerHelper.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/ControllerHelper.cs:1]: Put in AuthChainHelpers.cs instead?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs [hunk 6]
```diff
@@ -0,0 +1,735 @@
...
+                }
+
+                IHttpRequestAuthenticator authenticator = await this.authenticatorGetter;
+                HttpAuthResult authResult = await authenticator.AuthenticateAsync(deviceId, Option.None<string>(), this.HttpContext);
+
+                if (!authResult.Authenticated)
+                {
+                    Events.AuthenticateFail(deviceId);
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs:69]: output error message of the authResult? Applies to other instances too

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs [hunk 7]
```diff
@@ -0,0 +1,735 @@
...
+                await this.SendResponseAsync(responseMessage);
+                Events.CompleteRequest(edgeDeviceId, targetDeviceId, requestData.ModuleId, authChain);
+            }
+            catch (Exception ex)
+            {
+                await this.SendResponseAsync(HttpStatusCode.InternalServerError, this.FormatErrorResponseMessage(ex));
+            }
+        }
+
+        async Task<(bool, string)> AuthorizeActorAsync(IDeviceScopeIdentitiesCache identitiesCache, string actorDeviceId, string actorModuleId, string targetId)
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs:547]: >AuthorizeActorAsync [](start = 35, length = 19)

seems identitical to the one in DeviceScopeController, maybe put it in HttpRequestAuthenticator to share across both classes.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs [hunk 8]
```diff
@@ -0,0 +1,735 @@
...
+                IEdgeHub edgeHub = await this.edgeHubGetter;
+                string edgeDeviceId = edgeHub.GetEdgeDeviceId();
+                IDeviceScopeIdentitiesCache identitiesCache = edgeHub.GetDeviceScopeIdentitiesCache();
+
+                (bool authorized, string authChain) = await this.AuthorizeActorAsync(identitiesCache, edgeDeviceId, Constants.EdgeHubModuleId, deviceId);
+                if (!authorized)
+                {
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs:82]: this is probably unnecessary, AuthZ is only needed a child Edge acting OnBehalfOf a grandchild.  For the legacy endpoint, the current-Edge is always authorized to act OnBehalfOf of any child that connects to us.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs (hunk 6 above)
Comment (philipktlin) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs:69]: Current Registry API will return 401 status code with empty body.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeTokenAuthenticator.cs [hunk 9]
```diff
@@ -97,7 +97,7 @@ internal bool ValidateAudienceIds(string audienceDeviceId, Option<string> audien
             if (!audienceModuleId.HasValue)
             {
                 // Token is for a device
-                if (identity is IDeviceIdentity deviceIdentity && deviceIdentity.DeviceId != audienceDeviceId)
+                if (identity is IDeviceIdentity deviceIdentity && !string.Equals(deviceIdentity.DeviceId, audienceDeviceId, StringComparison.OrdinalIgnoreCase))
```
Comment (philipktlin) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeTokenAuthenticator.cs:100]: TODO: need to confirm with Richard

**Pull request closed without merging**